pkg/core/secrets/manager/validator.go

```diff
@@ -0,0 +1,72 @@
+package manager
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	mesh_proto "github.com/Kong/kuma/api/mesh/v1alpha1"
+	"github.com/Kong/kuma/pkg/core/ca"
+	mesh_core "github.com/Kong/kuma/pkg/core/resources/apis/mesh"
+	"github.com/Kong/kuma/pkg/core/resources/model"
+	core_store "github.com/Kong/kuma/pkg/core/resources/store"
+	"github.com/Kong/kuma/pkg/core/validators"
+)
+
+// MeshAccessor is used instead of MeshManager, otherwise we would have cyclic dependency.
+// MeshManager depends on SecretManager which depends on SecretValidator which would depend on ResourceManager
+type MeshAccessor = func(context.Context, model.Resource, ...core_store.GetOptionsFunc) error
+
+type SecretValidator interface {
+	ValidateDelete(ctx context.Context, secretName string, secretMesh string) error
+}
+
+func NewSecretValidator(caManagers ca.Managers, meshAccessor MeshAccessor) SecretValidator {
+	return &secretValidator{
+		caManagers:   caManagers,
+		meshAccessor: meshAccessor,
+	}
+}
+
+type secretValidator struct {
+	caManagers   ca.Managers
+	meshAccessor MeshAccessor
+}
+
+func (s *secretValidator) ValidateDelete(ctx context.Context, secretName string, secretMesh string) error {
+	mesh := &mesh_core.MeshResource{}
+	err := s.meshAccessor(ctx, mesh, core_store.GetByKey(secretMesh, secretMesh))
+	if err != nil {
+		return err
+	}
+
+	var verr validators.ValidationError
+	for _, backend := range mesh.Spec.GetMtls().GetBackends() {
+		used, err := s.secretUsedByMTLSBackend(secretName, mesh.GetMeta().GetName(), backend)
+		if err != nil {
+			return err
```

Comment:
If we have an error here then we won't delete that secret. But that error might occur because of broken Mesh format. Is it okay to leave a secret in the cluster in such case?

Reply:
That's a good point. I changed to only log the error